### PR TITLE
Refactored code from the console branch

### PR DESCRIPTION
**Pull request by JCash**
Changes to Lua:

* The input event for gamepad connected/disconnected now has a `user_id` for each gamepad.

* Added `sys.get_host_path(somepath)` for easier access on the host machine from the device (e.g. tests, logs etc)

Changes to dmSdk:

* Added dmScript::CheckTable, dmScript::CheckTableSize and dmScript::PushTable

* Added `dmsdk/dlib/sys.h`, to help with file I/O. (e.g. dmSys::Stat(), dmSys::Exists(), dmSys::Rename() and dmSys::Unlink())

* Added `dmsdk/dlib/memory.h` with dmMemory::AlignedMalloc, dmMemory::AlignedFree

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: engine/engine/src/engine.cpp [hunk 1]
```diff
@@ -775,7 +780,7 @@ namespace dmEngine
             char application_support_path[DMPATH_MAX_PATH];
             char application_support_log_path[DMPATH_MAX_PATH];
             const char* logs_dir = dmConfigFile::GetString(engine->m_Config, "project.title_as_file_name", "defoldlogs");
-            if (dmSys::GetApplicationSavePath(logs_dir, application_support_path, sizeof(application_support_path)) == dmSys::RESULT_OK)
+            if (dmSys::GetApplicationSupportPath(logs_dir, application_support_path, sizeof(application_support_path)) == dmSys::RESULT_OK)
```
Comment (JCash) [inline engine/engine/src/engine.cpp:783]: On consoles, a save path is connected to a user, which isn't really the case here.

File: engine/gamesys/content/materials/gui.vp [hunk 2]
```diff
@@ -1,5 +1,4 @@
 uniform mediump mat4 view_proj;
-uniform mediump mat4 world;
```
Comment (JCash) [inline engine/gamesys/content/materials/gui.vp:2]: Unused, and it failed loading properly

File: engine/gamesys/src/gamesys/test/resource/create_texture.script [hunk 3]
```diff
@@ -21,141 +21,143 @@ local function assert_error(func)
...
-	end)
+    assert_error(function()
+        resource.release("/not_found")
+    end)
 end
 
 function test_compressed()
-	local f     = io.open("src/gamesys/test/resource/blank.basis", "rb")
+    local host_path = sys.get_host_path("src/gamesys/test/resource/blank.basis")
+    local f     = io.open(host_path, "rb")
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/resource/create_texture.script:87]: Uses `sys.get_host_path()` to read from the host.

File: engine/gamesys/src/gamesys/test/resource/create_texture.script [hunk 4]
```diff
@@ -21,141 +21,143 @@ local function assert_error(func)
 end
 
 function init(self)
-	self.update_counter = 0
+    self.update_counter = 0
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/resource/create_texture.script:24]: Mostly space indentation

File: engine/gamesys/src/gamesys/test/test_gamesys.cpp [hunk 5]
```diff
@@ -710,9 +695,6 @@ TEST_F(SoundTest, UpdateSoundResource)
 
     dmGameSystem::InitializeScriptLibs(scriptlibcontext);
 
-    // Since the script uses "io.open()" we want to override the path
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/test_gamesys.cpp:713]: Previous test hack was to monkey patch the Lua function

File: engine/resource/wscript [hunk 6]
```diff
@@ -50,26 +51,63 @@ def build(bld):
...
         except urllib.error.URLError:
             print('Waiting for HTTP testserver to start...')
             sys.stdout.flush()
             time.sleep(0.5)
 
-    waf_dynamo.run_tests(test_context, valgrind = True)
+def _remove_config_file(path):
+    os.remove(path)
+
+def server(self):
```
Comment (JCash) [inline engine/resource/wscript:85]: When debugging the unit tests, you can now start a standalone server that it can use, with `waf server`

**Review by Jhonnyg — DISMISSED** [review]
A bit of a mental load, so I'm just gonna go ahead and approve this :)

**Pull request merged**